Adding initial code for PNC Interface

thescouser89 opened the pull request:
[DA-7]

The PNCInterface code will allow DA to talk with PNC.

We are currently using the Resteasy client to do REST calls to PNC.

Properties need to be defined in: `communication/src/main/resources/credentials.properties`
to setup the PNC server location, and keycloak properties.

This is an initial PR to see if I got most things right.

Things I'm unclear/not happy right now:
* How should we define the keycloak and pnc settings?
  - right now it's set in a properties file
  - should we use BASH variables instead?

* Should I allow the user of the `getPNCInstance()` method to specify the location of PNC she wants to use?

In the very near future I'll add the ability to ask PNC to add a new product/project etc. For now I just wanted to get some initial comments/review before moving further.

slaskawi commented:
Regarding to property file vs BASH properties - I would prefer using a JSON file just like we do in Orchestrator. This way it would be unified.

slaskawi commented on communication/pom.xml:
```diff
@@ -1,9 +1,39 @@
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
```
Is this dep needed? The connectivity should be provided by the RestEasy client.

thescouser89 replied:
Heh you are right. This is not needed. Somehow I thought I had to add those to get stuff working cause my IDE wasn't resolving the dependencies.

slaskawi commented on communication/src/main/java/PNCInterface.java:
```diff
@@ -0,0 +1,31 @@
+import javax.ws.rs.core.MediaType;
+import java.util.List;
+
+public interface PNCInterface {
```
In short-term this is OK, but I would consider small refactoring in Orchestrator and providing an artifact with the contract.

thescouser89 replied:
@slaskawi Do you mean the `Configuration` file in the NCL codebase?
If so, do you think I should also adopt the "@class" and "@module-config" JSON format?

slaskawi commented on communication/src/main/java/org/jboss/da/communcation/pnc/ReadResource.java:
```diff
@@ -0,0 +1,25 @@
+            prop.load(inputStream);
+            return prop.getProperty(key);
+        } catch (IOException e) {
+            e.printStackTrace();
```
No properties = fatal error IMO.

jbartece commented:
Maybe it would be good to create a new module for this type of methods. It could be named "common" or something like that

jbartece commented:
@thescouser89 btw. It is be good to add the JIRA id in the commit msg

janinko commented on communication/src/main/java/PNCInterface.java:
```diff
@@ -0,0 +1,31 @@
+public interface PNCInterface {
+
+    @GET
+    @Path("pnc-rest/rest/build-configurations")
```
Shuldn't the "pnc-rest/rest/" part be in target?

thescouser89 replied:
So, the target will become "<pnc-website>/pnc-rest/rest"?

I don't know. I kinda like that I just have to specify the main url in the target and not have to remember to add the `/pnc-rest/rest`. What does @jbartece or @slaskawi think?

janinko commented on communication/src/main/java/org/jboss/da/communcation/pnc/authentication/PNCAuthentication.java:
```diff
@@ -0,0 +1,126 @@
+
+        String refreshToken = "";
+        String accessToken = "";
+        try {
```
Can try-with-resources be used?

thescouser89 replied:
Weeeeeeeeeeee Java-7 stuff. I've been in Java-6 land for far too long :( will refactor!

janinko commented on communication/src/main/java/org/jboss/da/communcation/pnc/authentication/PNCAuthFilter.java:
```diff
@@ -0,0 +1,25 @@
+     * @throws IOException
+     */
+    public void filter(ClientRequestContext requestContext) throws IOException {
+        String token = PNCAuthentication.authenticate();
```
How long the token is valid? Do we need to authenticate again for every request or can we reuse the token?

pslegr replied:
It depends on the server configuration for tokens validity. As default Keycloak tokens are valid for 5 minutes. It is possible for sure to reuse the access_token, but then you would need to implement the logic to be prepared getting from REST service endpoint response as: "access token expired" and then re-authenticate and try again to access REST service. Of course, refresh tokens are also possibility for getting fresh access_tokens, but it would require enen more additional coding to handle the refreshing tokens.

janinko commented on communication/src/main/java/PNC.java:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @return instance of PNCInterface
+     */
+    public static PNCInterface getPNCInstance(boolean isAuthenticated) {
```
For usage in container environment, I think this should be producer method (@Produces).

thescouser89 replied:
So that `PNCInterface` can be `@Injected` ? Gotcha!

thescouser89 replied:
I'm still thinking in non-containerized Java land :) Thanks for all the input! Will keep that in mind for my next PR!

janinko commented on communication/src/main/java/org/jboss/da/communcation/pnc/ReadResource.java:
```diff
@@ -0,0 +1,25 @@
+ */
+public class ReadResource {
+
+    public static String getResource(String key) {
```
IMHO this should be bean and Properties loaded in @PostConstruct method.
I agree with slaskawi that when the properties file doesn't exist, there should be fatal error.

thescouser89 replied:
I'll probably remove that file, or rename it to "Configuration". So this class would get PostConstructed with all the properties in it, and then we just have to query it?

jbartece commented:
Rebase to master please. There are significant changes in the POM files. Probably you won't be able to build it, because you don't use the managed dependency versions ;-)

thescouser89 commented:
What can I do to get the `@Inject` annotation to work with JUnit tests?

I've refactored the configuration code and moved it to common. Right now it's really simple, and I'm not sure if it should become as fancy as the one in PNC. Any comments welcome :)

janinko commented on communication/src/main/java/org/jboss/da/communcation/pnc/PNCProducer.java:
```diff
@@ -0,0 +1,29 @@
+public class PNCProducer {
+    Configuration config = new Configuration();
+    /**
+     * Static factory to create an instance of org.jboss.da.communcation.pnc.PNCInterface
```
It's not static anymore.

janinko commented on common/src/main/java/org/jboss/da/common/json/DAConfig.java:
```diff
@@ -0,0 +1,36 @@
+package org.jboss.da.common.json;
+
+public class DAConfig implements AutoCloseable {
```
Why add AutoCloseable? There is nothing to close.

janinko commented on communication/src/main/java/org/jboss/da/communcation/pnc/authentication/PNCAuthentication.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    public String authenticate() {
+        try(DAConfig conf = config.getConfig()) {
```
There is no point for using try-with-resources, as DAConfig shouldn't be closable.

janinko commented on communication/src/main/java/org/jboss/da/communcation/pnc/PNCProducer.java:
```diff
@@ -0,0 +1,36 @@
+    public PNC getPNCInstance() {
+        ResteasyClient client = new ResteasyClientBuilder().build();
+
+        try (DAConfig conf = config.getConfig()) {
```
There is no point for using try-with-resources, as DAConfig shouldn't be closable.

jomrazek commented:
:+1:

janinko commented on common/src/main/java/org/jboss/da/common/util/Configuration.java:
```diff
@@ -0,0 +1,44 @@
+            configFileName = "da-config.json";
+        }
+        // TODO: refactor with proper logger
+        System.out.println("Loading configuration from file: " + configFileName);
```
So change it to logger or remove it altogether.

janinko commented:
After rebase and ``System.out.println`` removal :+1:

thescouser89 commented:
See #20 

Closing this one since the branch got a little bit messed up :)
